### ISSUE TITLE
Improve rockstor.service re robustness

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2657]
Thanks to @FroggyFlox for highlighting this issue: We still host a Python based services manager (supervisord) within our OS based services manager of systemd. The mid term goal is to drop our use of supervisord entirely - as a project we pre-date systemd and originally used supervisord exclusively. However our current supervisord hosting within our rockstor.service has recently had a short-comming surfaced while address another bug where we intended to stop older package services duing Web-UI update but in fact did not:

"Improve Web-UI update re systemd service management #2651" #2656
rpmbuild counterparts: "Ensure systemd services are stopped during upgrade ..." 
- https://github.com/rockstor/rockstor-rpmbuild/pull/44
- https://github.com/rockstor/rockstor-rpmbuild/issues/43

While addressing the above two issue/pr sets we see our failing; to be addressed under this issue: that of a long time-out/fail-through suspected as originating from our layered, and insufficiently configured rockstor.service/supervisord interface.
Excerpt from the former PR listed above, comment: https://github.com/rockstor/rockstor-rpmbuild/pull/44#issuecomment-1680906437

```
Aug 16 16:57:33 rleap15-5 systemd[1]: Stopping Rockstor startup script...
Aug 16 16:57:33 rleap15-5 poetry[13455]: 2023-08-16 16:57:33,731 INFO exited: data-collector (terminated by SIGTERM; not expected)
Aug 16 16:57:33 rleap15-5 poetry[13455]: 2023-08-16 16:57:33,731 WARN received SIGTERM indicating exit request
Aug 16 16:57:33 rleap15-5 poetry[13455]: 2023-08-16 16:57:33,731 INFO waiting for gunicorn, ztask-daemon to die
Aug 16 16:57:33 rleap15-5 poetry[13455]: 2023-08-16 16:57:33,931 INFO stopped: ztask-daemon (exit status 0)
Aug 16 16:57:36 rleap15-5 poetry[13455]: 2023-08-16 16:57:36,936 INFO waiting for gunicorn to die
Aug 16 16:57:39 rleap15-5 poetry[13455]: 2023-08-16 16:57:39,940 INFO waiting for gunicorn to die
Aug 16 16:57:42 rleap15-5 poetry[13455]: 2023-08-16 16:57:42,945 INFO waiting for gunicorn to die
Aug 16 16:57:45 rleap15-5 poetry[13455]: 2023-08-16 16:57:45,949 WARN killing 'gunicorn' (13471) with SIGKILL
Aug 16 16:57:45 rleap15-5 poetry[13455]: 2023-08-16 16:57:45,949 INFO waiting for gunicorn to die
Aug 16 16:57:46 rleap15-5 poetry[13455]: 2023-08-16 16:57:46,951 INFO stopped: gunicorn (terminated by SIGKILL)
Aug 16 16:58:46 rleap15-5 systemd[1]: rockstor.service: Deactivated successfully.
Aug 16 16:58:46 rleap15-5 systemd[1]: Stopped Rockstor startup script.
```

Above context provides the reproducer (Web-UI update with the proposed changes referenced earlier here.

It is proposed that we explore configuration improvement within our rockstor.service to assist with its management of supervisord in the context we require.

**Comment by phillxnet** [comment]
The counterpart supervisord logs associated with the cited comment excerpt:

# Supervisord
> less /opt/rockstor/var/log/supervisord.log
```
2023-08-16 16:57:33,731 INFO exited: data-collector (terminated by SIGTERM; not expected)
2023-08-16 16:57:33,731 WARN received SIGTERM indicating exit request
2023-08-16 16:57:33,731 INFO waiting for gunicorn, ztask-daemon to die
2023-08-16 16:57:33,931 INFO stopped: ztask-daemon (exit status 0)
2023-08-16 16:57:36,936 INFO waiting for gunicorn to die
2023-08-16 16:57:39,940 INFO waiting for gunicorn to die
2023-08-16 16:57:42,945 INFO waiting for gunicorn to die
2023-08-16 16:57:45,949 WARN killing 'gunicorn' (13471) with SIGKILL
2023-08-16 16:57:45,949 INFO waiting for gunicorn to die
2023-08-16 16:57:46,951 INFO stopped: gunicorn (terminated by SIGKILL)
```

## Huey (ztask-daemon)
Bracketed name, and rockstor-core service name reference is for historical reasons: smoother transition from the prior scheduler of ztaskd to Huey.
>  less /opt/rockstor/var/log/huey.log
```
[2023-08-16 16:57:33,717] INFO:huey.consumer:MainThread:Received SIGTERM
[2023-08-16 16:57:33,731] INFO:huey.consumer:MainThread:Received SIGTERM
[2023-08-16 16:57:33,789] INFO:huey.consumer:MainThread:Shutting down
[2023-08-16 16:57:33,789] INFO:huey.consumer:MainThread:Consumer exiting.
```

## Gunicorn
> less /opt/rockstor/var/log/gunicorn.log
```
[2023-08-16 16:57:33 +0100] [13471] [INFO] Handling signal: term
[2023-08-16 16:57:33 +0100] [13481] [INFO] Worker exiting (pid: 13481)
[2023-08-16 16:58:46 +0100] [13482] [INFO] Worker exiting (pid: 13482
```

**Comment by phillxnet** [comment]
The  invocation of supervisord and it's charges via rockstor.service results in the following:
> ps axjf

```
 1 13471 13471 13471 ?           -1 Ss       0   0:04 /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/supervisord -c /opt/rockstor/etc/supervisord.conf
13471 13487 13487 13471 ?           -1 S        0   0:02  \_ /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/gunicorn --bind=127.0.0.1:8000 --pid=/run/gunicorn.pid --workers=2 --log-file=/opt/rockstor/var/log/gunicorn.log --pythonpath=/opt/rockstor/src/rockstor --timeout=120 --graceful-timeout=120 wsgi:application
13487 13495 13487 13471 ?           -1 S        0   0:00  |   \_ /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/gunicorn --bind=127.0.0.1:8000 --pid=/run/gunicorn.pid --workers=2 --log-file=/opt/rockstor/var/log/gunicorn.log --pythonpath=/opt/rockstor/src/rockstor --timeout=120 --graceful-timeout=120 wsgi:application
13487 13498 13487 13471 ?           -1 S        0   0:01  |   \_ /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/gunicorn --bind=127.0.0.1:8000 --pid=/run/gunicorn.pid --workers=2 --log-file=/opt/rockstor/var/log/gunicorn.log --pythonpath=/opt/rockstor/src/rockstor --timeout=120 --graceful-timeout=120 wsgi:application
13471 13488 13488 13471 ?           -1 S        0   0:01  \_ /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/data-collector
13471 13489 13489 13471 ?           -1 Sl       0   0:21  \_ /opt/rockstor/.venv/bin/python /opt/rockstor/.venv/bin/django-admin run_huey --workers 2 --worker-type thread --logfile /opt/rockstor/var/log/huey.log
```

**Comment by phillxnet** [comment]
From our supervisord configuration we have:
```
nodaemon=true                ; (start in foreground if true;default false)
```
so we are are currently running it in foreground mode, not deamon mode which would apparently need our systemd service to be `Type=forking`.

**Comment by phillxnet** [comment]
Proposed ExecStop
> /root/.local/bin/poetry run supervisorctl shutdown
```
2023-08-17 14:25:54,661 INFO waiting for gunicorn, data-collector, ztask-daemon to die
2023-08-17 14:25:54,858 INFO stopped: ztask-daemon (exit status 0)
2023-08-17 14:25:55,863 INFO stopped: data-collector (terminated by SIGTERM)
2023-08-17 14:25:57,095 INFO stopped: gunicorn (exit status 0)

```
Proposed ExecReload
> /root/.local/bin/poetry run supervisorctl reload
Works only if the service is running:
```
Aug 17 14:33:37 rleap15-5 poetry[15929]: 2023-08-17 14:33:37,276 INFO waiting for gunicorn, data-collector, ztask-daemon to die
Aug 17 14:33:37 rleap15-5 poetry[15929]: 2023-08-17 14:33:37,473 INFO stopped: ztask-daemon (exit status 0)
Aug 17 14:33:38 rleap15-5 poetry[15929]: 2023-08-17 14:33:38,478 INFO stopped: data-collector (terminated by SIGTERM)
Aug 17 14:33:39 rleap15-5 poetry[15929]: 2023-08-17 14:33:39,710 INFO stopped: gunicorn (exit status 0)
Aug 17 14:33:39 rleap15-5 poetry[15929]: 2023-08-17 14:33:39,718 INFO Set uid to user 0 succeeded
Aug 17 14:33:39 rleap15-5 poetry[15929]: 2023-08-17 14:33:39,719 INFO RPC interface 'supervisor' initialized
Aug 17 14:33:39 rleap15-5 poetry[15929]: 2023-08-17 14:33:39,719 INFO supervisord started with pid 15929
Aug 17 14:33:40 rleap15-5 poetry[15929]: 2023-08-17 14:33:40,722 INFO spawned: 'gunicorn' with pid 15981
Aug 17 14:33:40 rleap15-5 poetry[15929]: 2023-08-17 14:33:40,725 INFO spawned: 'data-collector' with pid 15982
Aug 17 14:33:40 rleap15-5 poetry[15929]: 2023-08-17 14:33:40,728 INFO spawned: 'ztask-daemon' with pid 15983
Aug 17 14:33:42 rleap15-5 poetry[15929]: 2023-08-17 14:33:42,731 INFO success: data-collector entered RUNNING state, process has stayed up for > than 2 seconds (startsecs)
Aug 17 14:33:45 rleap15-5 poetry[15929]: 2023-08-17 14:33:45,735 INFO success: gunicorn entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
Aug 17 14:33:50 rleap15-5 poetry[15929]: 2023-08-17 14:33:50,743 INFO success: ztask-daemon entered RUNNING state, process has stayed up for > than 10 seconds (startsecs)
```

**Comment by phillxnet** [comment]
We have an upstream Tumbleweed reference with systemd service files example, however, this example implementation pertains to a deamon mode configuration: which we do not. Noting for context and potential later reference:
https://build.opensuse.org/package/show/openSUSE:Factory/supervisor

**Comment by phillxnet** [comment]
When using KillMode=process, along with the above ExecStop we get the following when `systemctl stop rockstor` is executed by root from the cli:
```
Aug 17 15:20:14 rleap15-5 systemd[1]: Stopping Rockstor startup script...
Aug 17 15:20:15 rleap15-5 poetry[23758]: 2023-08-17 15:20:15,896 INFO waiting for gunicorn, data-collector, ztask-daemon to die
Aug 17 15:20:15 rleap15-5 poetry[23901]: Shut down
Aug 17 15:20:16 rleap15-5 poetry[23758]: 2023-08-17 15:20:16,051 INFO stopped: ztask-daemon (exit status 0)
Aug 17 15:20:16 rleap15-5 poetry[23758]: 2023-08-17 15:20:16,051 WARN received SIGTERM indicating exit request
Aug 17 15:20:17 rleap15-5 poetry[23758]: 2023-08-17 15:20:17,057 INFO stopped: data-collector (terminated by SIGTERM)
Aug 17 15:20:18 rleap15-5 poetry[23758]: 2023-08-17 15:20:18,287 INFO stopped: gunicorn (exit status 0)
Aug 17 15:20:18 rleap15-5 systemd[1]: rockstor.service: Deactivated successfully.
Aug 17 15:20:18 rleap15-5 systemd[1]: Stopped Rockstor startup script.
```

**Comment by phillxnet** [comment]
With the proposed changes covered to date in this issues comments applied to an outgoing instance of rockstor (run by the incoming rpm to stop prior instance servcies) we have the following result from a Web-UI update:

## journalctl -f
```
Aug 17 15:40:22 rleap15-5 systemd[1]: Stopping Rockstor startup script...
Aug 17 15:40:23 rleap15-5 poetry[13569]: 2023-08-17 15:40:23,655 INFO waiting for gunicorn, data-collector, ztask-daemon to die
Aug 17 15:40:23 rleap15-5 poetry[16272]: Shut down
Aug 17 15:40:23 rleap15-5 poetry[13569]: 2023-08-17 15:40:23,803 INFO stopped: ztask-daemon (exit status 0)
Aug 17 15:40:23 rleap15-5 poetry[13569]: 2023-08-17 15:40:23,803 WARN received SIGTERM indicating exit request
Aug 17 15:40:24 rleap15-5 poetry[13569]: 2023-08-17 15:40:24,813 INFO stopped: data-collector (terminated by SIGTERM)
Aug 17 15:40:26 rleap15-5 poetry[13569]: 2023-08-17 15:40:26,816 INFO waiting for gunicorn to die
Aug 17 15:40:29 rleap15-5 poetry[13569]: 2023-08-17 15:40:29,821 INFO waiting for gunicorn to die
Aug 17 15:40:32 rleap15-5 poetry[13569]: 2023-08-17 15:40:32,825 INFO waiting for gunicorn to die
Aug 17 15:40:34 rleap15-5 poetry[13569]: 2023-08-17 15:40:34,828 WARN killing 'gunicorn' (13586) with SIGKILL
Aug 17 15:40:35 rleap15-5 poetry[13569]: 2023-08-17 15:40:35,830 INFO stopped: gunicorn (terminated by SIGKILL)
Aug 17 15:40:35 rleap15-5 systemd[1]: rockstor.service: Deactivated successfully.
Aug 17 15:40:35 rleap15-5 systemd[1]: rockstor.service: Unit process 13592 (gunicorn) remains running after unit stopped.
Aug 17 15:40:35 rleap15-5 systemd[1]: Stopped Rockstor startup script.
```

## tail -f /opt/rockstor/var/log/supervisord.log

```
2023-08-17 15:40:23,655 INFO waiting for gunicorn, data-collector, ztask-daemon to die
2023-08-17 15:40:23,803 INFO stopped: ztask-daemon (exit status 0)
2023-08-17 15:40:23,803 WARN received SIGTERM indicating exit request
2023-08-17 15:40:24,813 INFO stopped: data-collector (terminated by SIGTERM)
2023-08-17 15:40:26,816 INFO waiting for gunicorn to die
2023-08-17 15:40:29,821 INFO waiting for gunicorn to die
2023-08-17 15:40:32,825 INFO waiting for gunicorn to die
2023-08-17 15:40:34,828 WARN killing 'gunicorn' (13586) with SIGKILL
2023-08-17 15:40:35,830 INFO stopped: gunicorn (terminated by SIGKILL)
```

So we have an apparently successful services shutdown executed by supervisord, but instructed/requested by systemd.

## Logged stop times rockstor.service
As run via AT update script (Web-UI update), as per our issue reproducer.
- Before proposed changes: 1 min 13 seconds.
- After proposed changes: 13 seconds.

**Comment by phillxnet** [comment]
Pull request to follow shortly.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2658
